test/functoria/test_graph.ml

```diff
@@ -10,34 +10,36 @@ let z = Device.v "Bar" job
 
 let apply f d =
   let id = Device.id d in
-  let g = Graph.create (of_device d) in
+  let g = Impl.abstract (of_device d) in
+  let g = Impl.eval ~context:Key.empty_context g in
   let v =
-    match
-      Graph.find_all g (function Graph.Dev d -> Device.id d = id | _ -> false)
-    with
-    | [ x ] -> x
-    | _ -> assert false
+    let f x l =
+      match x with
+      | Device.Graph.D { dev; _ } when Device.id dev = id -> f x :: l
+      | _ -> l
+    in
+    match Device.Graph.fold f g [] with [ x ] -> x | _ -> assert false
   in
-  f v
+  v
 
-let var_name x = apply Graph.var_name x
+let var_name x = apply Device.Graph.var_name x
 
-let impl_name x = apply Graph.impl_name x
+let impl_name x = apply Device.Graph.impl_name x
 
 let id () = Scanf.sscanf (var_name zero) "z__%d" (fun i -> i)
 
-let ident s i x = Fmt.strf "%s__%d" s (i + x)
+let ident s i = Fmt.strf "%s__%d" s i
 
 let test_var_name () =
   let id = id () in
-  Alcotest.(check string) "x" (ident "foo_bar" id 1) (var_name x);
-  Alcotest.(check string) "y" (ident "x_y" id 2) (var_name y);
-  Alcotest.(check string) "z" (ident "bar" id 3) (var_name z)
+  Alcotest.(check string) "x" (ident "foo_bar" id) (var_name x);
```

Comment:
I am confused: why is this different now? How do we ensure devices have unique names (and variable names)?

Reply:
Okay, so the point of the test was indeed rendered a bit moot. I modified it partially: Now, we create a full implementation, pass it through the eval function, ask for a topo sort and tests the IDs on that. What do you think of that ?